Handle filesystem latency when creating `RUN_COMPLETE` file

hwikle-lanl opened the pull request:
Code review checklist:

- [ ] Code is generally sensical and well commented
- [ ] Variable/function names all telegraph their purpose and contents
- [ ] Functions/classes have useful doc strings
- [ ] Function arguments are typed
- [ ] Added/modified unit tests to cover changes.
- [ ] New features have documentation added to the docs.
- [ ] New features and backwards compatibility breaks are noted in the RELEASE.md

hwikle-lanl committed:
Catch unhandled TimeoutError and log status

hwikle-lanl committed:
Increase RUN_COMPLETE timeout to 30 seconds

File system quirks mean that the RUN_COMPLETE file may not yet exist
on the file system after it is created. Since the file is used by
the same process relatively soon after creation, this can occassionally
cause a FileNotFoundError. It was therefore necessary to wait for
the file to be synced to the file system, and to impose a timeout.
The previous timeout value was too short and resulted in TimeoutErrors
relatively frequently. This commit increases the timeout value from
2 seconds to 30 seconds, per the suggestion of Paul Ferrell.

Paul-Ferrell reviewed (commented):
I really don't like throwing an exception here. 
I think we should try harder to force a cache update. Instead of just doing an `exists`, you might try tossing in a `path.parent.iterdir()`. Apparently opening the parent directory should cause the directory cache to refresh.

If that doesn't seem to be working, drop a symlink to the temp file location where you were going to move the original file, and walk away instead of throwing an exception.

Also, CREATION_ERROR is explicitly for test creation problems. I'd create a new 'WARNING' label for cases like this.

hwikle-lanl committed:
Add fallback if RUN_COMPLETE is not present

The idiosyncrocies of NFS means that when the RUN_COMPLETE file is
written, it may not appear on the local file system, which results
in a FileNotFoundError when that file is accessed. This commit adds
a fallback when the file does not exist on the local file system, in
the process reverting the previous solution which involved waiting
and eventually raising a timeout.

The fallback consists of the following steps:

1. Create the file as normal, and check whether it exists.

2. If it does not exist, attempt to force an NFS cache reset by listing
the contents of the parent directory.

3. If the file still does not exist, create a symlink to the location
where the file is expected eventually to be.

hwikle-lanl committed:
Add unit test for missing RUN_COMPLETE file fallback

hwikle-lanl committed:
Fix failing style tests

hwikle-lanl committed:
Fix error

smehta99 reviewed (approved):
LGTM
